# Practice Questions.

1) How many shows were released in each year?

2) What is the most common rating?

3) What's the average duration of shows per country?

4) Can you find the genre with the highest total views?

5) Filter shows with duration greater than 50 minutes and views > 1M.

6) Which countries have more than one show?

7) Add a new column: view_category → "Hit" if views > 1M, else "Average".

8) Find shows with titles containing the word "in".

In [2]:
import pandas as pd

data = {
    "show_id": [1, 2, 3, 4, 5, 6, 7, 8],
    "title": ["Stranger Things", "Money Heist", "The Crown", "Dark", "Wednesday", "Lupin", "Squid Game", "Emily in Paris"],
    "genre": ["Sci-Fi", "Action", "Drama", "Sci-Fi", "Comedy", "Crime", "Thriller", "Romance"],
    "release_year": [2016, 2017, 2016, 2017, 2022, 2021, 2021, 2020],
    "rating": ["TV-14", "TV-MA", "TV-14", "TV-MA", "TV-14", "TV-MA", "TV-MA", "TV-14"],
    "duration": ["45 min", "50 min", "58 min", "60 min", "47 min", "45 min", "55 min", "30 min"],
    "views": [1200000, 950000, 600000, 800000, 1100000, 500000, 1500000, 300000],
    "country": ["USA", "Spain", "UK", "Germany", "USA", "France", "South Korea", "France"]
}

df = pd.DataFrame(data)
print(df)

   show_id            title     genre  release_year rating duration    views  \
0        1  Stranger Things    Sci-Fi          2016  TV-14   45 min  1200000   
1        2      Money Heist    Action          2017  TV-MA   50 min   950000   
2        3        The Crown     Drama          2016  TV-14   58 min   600000   
3        4             Dark    Sci-Fi          2017  TV-MA   60 min   800000   
4        5        Wednesday    Comedy          2022  TV-14   47 min  1100000   
5        6            Lupin     Crime          2021  TV-MA   45 min   500000   
6        7       Squid Game  Thriller          2021  TV-MA   55 min  1500000   
7        8   Emily in Paris   Romance          2020  TV-14   30 min   300000   

       country  
0          USA  
1        Spain  
2           UK  
3      Germany  
4          USA  
5       France  
6  South Korea  
7       France  


# 🔍 1. Load and Inspect the Data.
Essential step to start the clean process when you started clean the data set.

import pandas as pd

df = pd.read_csv("your_dataset.csv")  # or read_excel(), read_json(), etc.
df.head()         # Preview first 5 rows
df.tail()         # Preview last 5 rows
df.shape          # (rows, columns)
df.info()         # Column types, non-null counts
df.describe()     # Stats summary (only for numeric columns)
df.columns        # List of column names



In [3]:
# display the first 5 rows.
df.head()

show_id            title   genre  release_year rating duration    views  \
0        1  Stranger Things  Sci-Fi          2016  TV-14   45 min  1200000   
1        2      Money Heist  Action          2017  TV-MA   50 min   950000   
2        3        The Crown   Drama          2016  TV-14   58 min   600000   
3        4             Dark  Sci-Fi          2017  TV-MA   60 min   800000   
4        5        Wednesday  Comedy          2022  TV-14   47 min  1100000   

   country  
0      USA  
1    Spain  
2       UK  
3  Germany  
4      USA

In [4]:


# display the last 5 rows.
df.tail()

show_id           title     genre  release_year rating duration    views  \
3        4            Dark    Sci-Fi          2017  TV-MA   60 min   800000   
4        5       Wednesday    Comedy          2022  TV-14   47 min  1100000   
5        6           Lupin     Crime          2021  TV-MA   45 min   500000   
6        7      Squid Game  Thriller          2021  TV-MA   55 min  1500000   
7        8  Emily in Paris   Romance          2020  TV-14   30 min   300000   

       country  
3      Germany  
4          USA  
5       France  
6  South Korea  
7       France

In [5]:
df.shape         # (rows, columns)

(8, 8)

In [6]:
df.info()         # Column types, non-null counts

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8 non-null      int64 
 1   title         8 non-null      object
 2   genre         8 non-null      object
 3   release_year  8 non-null      int64 
 4   rating        8 non-null      object
 5   duration      8 non-null      object
 6   views         8 non-null      int64 
 7   country       8 non-null      object
dtypes: int64(3), object(5)
memory usage: 644.0+ bytes


In [7]:
df.describe()     # Stats summary (only for numeric columns)

show_id  release_year         views
count  8.00000      8.000000  8.000000e+00
mean   4.50000   2018.750000  8.687500e+05
std    2.44949      2.492847  3.972562e+05
min    1.00000   2016.000000  3.000000e+05
25%    2.75000   2016.750000  5.750000e+05
50%    4.50000   2018.500000  8.750000e+05
75%    6.25000   2021.000000  1.125000e+06
max    8.00000   2022.000000  1.500000e+06


### 📊 **1. What’s the average number of views for my shows?**
> **Answer from `mean`:**  
The average view count across all shows is **868,750 views**.

---

### 🎯 **2. What’s my most-watched and least-watched show in terms of views?**
> **Answer from `max` and `min`:**  
- **Most-watched:** 1,500,000 views  
- **Least-watched:** 300,000 views  

---

### 📈 **3. How spread out are the views? Are there any big spikes?**
> **Answer from `std` (standard deviation):**  
The standard deviation is **~397k**, which means there is **significant variation**—some shows perform way better than others.

---

### 🕰️ **4. What’s the average release year of our catalog?**
> **Answer from `mean` of `release_year`:**  
The average release year is around **2018.75**, so most content is from **recent years**.

---

### 🔎 **5. Can you break views into performance tiers (low, average, high)?**
> **Answer using `25%`, `50%`, and `75%` (percentiles):**
- **25% of shows** have less than **575,000** views → Low Performers
- **50% (median)** is **875,000** → Typical Performance
- **Top 25%** have more than **1,125,000** views → High Performers

---

### 📉 **6. What are the oldest and newest shows in the catalog?**
> **Answer from `min` and `max` in `release_year`:**
- **Oldest Show:** Released in **2016**  
- **Newest Show:** Released in **2022**

---

### 🧠 Summary Insight You Can Offer Your Client:
> “Based on the current catalog, your top-performing shows get 3–5x more views than your lowest-performing ones. There’s a clear spike in high performance for newer releases post-2020. We can use these insights to optimize content strategy—by producing similar genre shows or analyzing what’s making these high performers successful.”

---



In [8]:
df.columns        # List of column names


Index(['show_id', 'title', 'genre', 'release_year', 'rating', 'duration',
       'views', 'country'],
      dtype='object')

```python
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')
```

This line is **cleaning and formatting column names** in a DataFrame (`df`) — a super important step in data cleaning, especially for professional projects.

---

### 🔍 What does it do?

You're chaining string operations to **standardize column names** in a format called **`snake_case`**, which is clean and Python-friendly.

Let’s break it into parts:

---

### 🧱 1. `df.columns`
This accesses the **column names** of your DataFrame.

Example before cleaning:

```python
[' Show ID ', 'Release Year', 'Views']
```

---

### 🧹 2. `.str.strip()`
This **removes leading and trailing spaces** from each column name.

```python
['Show ID', 'Release Year', 'Views']
```

---

### 🔡 3. `.str.lower()`
This converts **all letters to lowercase** — helps avoid case sensitivity bugs.

```python
['show id', 'release year', 'views']
```

---

### 🐍 4. `.str.replace(' ', '_')`
This replaces all **spaces** with **underscores**, making the names **snake_case** (preferred in Python).

```python
['show_id', 'release_year', 'views']
```

---

### ✅ Final Result:
```python
['show_id', 'release_year', 'views']
```

Now these column names are:
- Clean
- Consistent
- Easy to access in code (e.g., `df.show_id` or `df['show_id']`)

---

### 🧠 Why is this useful?

✅ Prevents bugs  
✅ Makes your code readable and consistent  
✅ Follows Python best practices (snake_case)

---

### 🎯 In short:
This line of code is a **one-liner magic wand** that:
- Trims spaces
- Lowers case
- Replaces spaces with underscores

Perfect for professional data analysis projects! 🚀

Let me know if you want a real example to test it out!

In [9]:
df.columns = df.columns.str.strip().str.upper().str.replace(' ', '_') # Snake_case and clean
df.head()

SHOW_ID            TITLE   GENRE  RELEASE_YEAR RATING DURATION    VIEWS  \
0        1  Stranger Things  Sci-Fi          2016  TV-14   45 min  1200000   
1        2      Money Heist  Action          2017  TV-MA   50 min   950000   
2        3        The Crown   Drama          2016  TV-14   58 min   600000   
3        4             Dark  Sci-Fi          2017  TV-MA   60 min   800000   
4        5        Wednesday  Comedy          2022  TV-14   47 min  1100000   

   COUNTRY  
0      USA  
1    Spain  
2       UK  
3  Germany  
4      USA

❓ 3. Check & Handle Missing Values
python
Copy code
df.isnull().sum()            # Count missing values per column
df.dropna()                  # Drop rows with any missing values
df.dropna(subset=['col'])    # Drop rows where 'col' is NaN
df.fillna(0)                 # Fill with 0
df['col'].fillna(df['col'].mean())  # Fill with column mean 

In [10]:
df.isnull().sum()  # counting Values

SHOW_ID         0
TITLE           0
GENRE           0
RELEASE_YEAR    0
RATING          0
DURATION        0
VIEWS           0
COUNTRY         0
dtype: int64

In [11]:
df.dropna()    # Drop rows with any missing values

SHOW_ID            TITLE     GENRE  RELEASE_YEAR RATING DURATION    VIEWS  \
0        1  Stranger Things    Sci-Fi          2016  TV-14   45 min  1200000   
1        2      Money Heist    Action          2017  TV-MA   50 min   950000   
2        3        The Crown     Drama          2016  TV-14   58 min   600000   
3        4             Dark    Sci-Fi          2017  TV-MA   60 min   800000   
4        5        Wednesday    Comedy          2022  TV-14   47 min  1100000   
5        6            Lupin     Crime          2021  TV-MA   45 min   500000   
6        7       Squid Game  Thriller          2021  TV-MA   55 min  1500000   
7        8   Emily in Paris   Romance          2020  TV-14   30 min   300000   

       COUNTRY  
0          USA  
1        Spain  
2           UK  
3      Germany  
4          USA  
5       France  
6  South Korea  
7       France

In [12]:
df.dropna(subset=['GENRE']) # Drop rows where 'col' is NaN

SHOW_ID            TITLE     GENRE  RELEASE_YEAR RATING DURATION    VIEWS  \
0        1  Stranger Things    Sci-Fi          2016  TV-14   45 min  1200000   
1        2      Money Heist    Action          2017  TV-MA   50 min   950000   
2        3        The Crown     Drama          2016  TV-14   58 min   600000   
3        4             Dark    Sci-Fi          2017  TV-MA   60 min   800000   
4        5        Wednesday    Comedy          2022  TV-14   47 min  1100000   
5        6            Lupin     Crime          2021  TV-MA   45 min   500000   
6        7       Squid Game  Thriller          2021  TV-MA   55 min  1500000   
7        8   Emily in Paris   Romance          2020  TV-14   30 min   300000   

       COUNTRY  
0          USA  
1        Spain  
2           UK  
3      Germany  
4          USA  
5       France  
6  South Korea  
7       France

In [13]:
df.fillna(0)                 # Fill with 0

SHOW_ID            TITLE     GENRE  RELEASE_YEAR RATING DURATION    VIEWS  \
0        1  Stranger Things    Sci-Fi          2016  TV-14   45 min  1200000   
1        2      Money Heist    Action          2017  TV-MA   50 min   950000   
2        3        The Crown     Drama          2016  TV-14   58 min   600000   
3        4             Dark    Sci-Fi          2017  TV-MA   60 min   800000   
4        5        Wednesday    Comedy          2022  TV-14   47 min  1100000   
5        6            Lupin     Crime          2021  TV-MA   45 min   500000   
6        7       Squid Game  Thriller          2021  TV-MA   55 min  1500000   
7        8   Emily in Paris   Romance          2020  TV-14   30 min   300000   

       COUNTRY  
0          USA  
1        Spain  
2           UK  
3      Germany  
4          USA  
5       France  
6  South Korea  
7       France

🔢 4. Fix Data Types
python
Copy code
df['date'] = pd.to_datetime(df['date'])             # Convert to datetime
df['price'] = df['price'].astype(float)             # Convert to float
df['category'] = df['category'].astype('category')  # Convert to category


✂️ 5. Remove Duplicates
python
Copy code
df.duplicated().sum()     # Count duplicate rows
df = df.drop_duplicates()  # Remove duplicate rows

In [14]:
df.duplicated().sum()

np.int64(0)

In [15]:
df = df.drop_duplicates()  # Remove duplicate rows

🎯 6. Filter & Clean Rows Based on Logic
python
Copy code
df = df[df['price'] > 0]                # Remove rows with invalid prices
df = df[df['status'].notnull()]         # Keep rows with valid status
df = df[df['country'].str.len() > 2]    # Filter out invalid

In [16]:
df = df[df['COUNTRY'].notnull()]

In [17]:
df = df[df['COUNTRY'].str.len() > 2]    # Filter out invalid

🔧 7. Clean Text Columns
python
Copy code
df['title'] = df['title'].str.strip().str.lower()           # Remove whitespace, lowercase
df['title'] = df['title'].str.replace('[^a-zA-Z0-9 ]', '')  

In [18]:
df['COUNTRY'] = df['COUNTRY'].str.strip().str.lower()    

In [19]:
df['TITLE'] = df['TITLE'].str.upper() 
df.head()

SHOW_ID            TITLE   GENRE  RELEASE_YEAR RATING DURATION    VIEWS  \
0        1  STRANGER THINGS  Sci-Fi          2016  TV-14   45 min  1200000   
1        2      MONEY HEIST  Action          2017  TV-MA   50 min   950000   
3        4             DARK  Sci-Fi          2017  TV-MA   60 min   800000   
4        5        WEDNESDAY  Comedy          2022  TV-14   47 min  1100000   
5        6            LUPIN   Crime          2021  TV-MA   45 min   500000   

   COUNTRY  
0      usa  
1    spain  
3  germany  
4      usa  
5   france

🧮 9. Create or Engineer New Columns
python
Copy code
df['duration_min'] = df['duration'].str.replace(' min', '').astype(int)
df['year'] = pd.cut(df['year'], bins=[2010, 2018, 2022], labels=['Old', 'New'])

In [20]:
if 'duration' in df.columns:
    df['duration_min'] = df['duration'].str.replace(' min', '', regex=False).astype(int)
df.head()

SHOW_ID            TITLE   GENRE  RELEASE_YEAR RATING DURATION    VIEWS  \
0        1  STRANGER THINGS  Sci-Fi          2016  TV-14   45 min  1200000   
1        2      MONEY HEIST  Action          2017  TV-MA   50 min   950000   
3        4             DARK  Sci-Fi          2017  TV-MA   60 min   800000   
4        5        WEDNESDAY  Comedy          2022  TV-14   47 min  1100000   
5        6            LUPIN   Crime          2021  TV-MA   45 min   500000   

   COUNTRY  
0      usa  
1    spain  
3  germany  
4      usa  
5   france

Practice Questions.¶
How many shows were released in each year?

In [21]:
# How many shows were released in each year?
print(df)

   SHOW_ID            TITLE     GENRE  RELEASE_YEAR RATING DURATION    VIEWS  \
0        1  STRANGER THINGS    Sci-Fi          2016  TV-14   45 min  1200000   
1        2      MONEY HEIST    Action          2017  TV-MA   50 min   950000   
3        4             DARK    Sci-Fi          2017  TV-MA   60 min   800000   
4        5        WEDNESDAY    Comedy          2022  TV-14   47 min  1100000   
5        6            LUPIN     Crime          2021  TV-MA   45 min   500000   
6        7       SQUID GAME  Thriller          2021  TV-MA   55 min  1500000   
7        8   EMILY IN PARIS   Romance          2020  TV-14   30 min   300000   

       COUNTRY  
0          usa  
1        spain  
3      germany  
4          usa  
5       france  
6  south korea  
7       france  


In [26]:
# Group by 'release_year' and count the number of shows per year.
show_per_year = df.groupby('RELEASE_YEAR').size()
print(show_per_year)

RELEASE_YEAR
2016    1
2017    2
2020    1
2021    2
2022    1
dtype: int64


This code will output the number of shows released each year by grouping the DataFrame based on the release_year column and counting the entries in each group.​

Alternatively, you can achieve the same result using the value_counts() method, which counts the occurrences of each unique value in a Series:

In [31]:
show_per_year = df.groupby('RELEASE_YEAR').value_counts().sort_index()
print(show_per_year)

RELEASE_YEAR  SHOW_ID  TITLE            GENRE     RATING  DURATION  VIEWS    COUNTRY    
2016          1        STRANGER THINGS  Sci-Fi    TV-14   45 min    1200000  usa            1
2017          2        MONEY HEIST      Action    TV-MA   50 min    950000   spain          1
              4        DARK             Sci-Fi    TV-MA   60 min    800000   germany        1
2020          8        EMILY IN PARIS   Romance   TV-14   30 min    300000   france         1
2021          6        LUPIN            Crime     TV-MA   45 min    500000   france         1
              7        SQUID GAME       Thriller  TV-MA   55 min    1500000  south korea    1
2022          5        WEDNESDAY        Comedy    TV-14   47 min    1100000  usa            1
Name: count, dtype: int64


In [34]:
# 1. Most Common Rating
# mode() returns all modes in case of a tie.
most_common_rating = df['RATING'].mode()
print(most_common_rating)


0    TV-MA
Name: RATING, dtype: object


2) Average Duration of Shows per Country
First, convert the 'duration' column to numeric values (in minutes):

In [47]:
# Converting the Duration columns to numeric values
df['duration_minutes'] = df['DURATION'].str.replace('min','').astype(int)

In [51]:
avg_duration = df.groupby('COUNTRY')['duration_minutes'].mean()
print(avg_duration)

248.5


3. Genre with the Highest Total Views
To find the genre with the highest total views:

In [55]:
genre_views = df.groupby('GENRE')['VIEWS'].sum()
top_genre = genre_views.idxmax()
print(top_genre)

Sci-Fi


In [58]:
# 4.Filter Shows with Duration > 50 Minutes and Views > 1 Million.
filtered_shows = df[(df['duration_minutes'] > 50) & (df['VIEWS'] > 1_000_000)]
print(filtered_shows[['TITLE','DURATION','VIEWS']])

        TITLE DURATION    VIEWS
6  SQUID GAME   55 min  1500000


In [60]:
# 5.Countries with More Than One Show
country_counts  = df['COUNTRY'].value_counts()
countries_with_multiple_shows = country_counts[country_counts > 1]
print(countries_with_multiple_shows)

COUNTRY
usa       2
france    2
Name: count, dtype: int64


In [62]:
# 6. Add a New Column: view_category
df['VIEW_CATEGORY'] = df['VIEWS'].apply(lambda x: 'HIT' if x > 1_000_000 else 'Average')
print(df[['TITLE','VIEWS','VIEW_CATEGORY']])


             TITLE    VIEWS VIEW_CATEGORY
0  STRANGER THINGS  1200000           HIT
1      MONEY HEIST   950000       Average
3             DARK   800000       Average
4        WEDNESDAY  1100000           HIT
5            LUPIN   500000       Average
6       SQUID GAME  1500000           HIT
7   EMILY IN PARIS   300000       Average


In [67]:
# 7 Find Shows with Titles Containing the Word "in"
shows_with_in = df[df['TITLE'].str.contains(r'\bdark\b', case=False, regex=True)]
print(shows_with_in[['TITLE']])

  TITLE
3  DARK


This line filters the DataFrame df to include only those rows where the title column contains the word "in" as a standalone word, not as part of another word. Here's a step-by-step explanation:

df['title']: Accesses the 'title' column of the DataFrame, which is a Series containing the titles of the shows.​

.str.contains(...): Applies a string method to each element in the Series to check if it contains a specified pattern.​

r'\bin\b': This is a regular expression pattern. Let's dissect it:

r'': The r prefix denotes a raw string in Python, which tells Python not to interpret backslashes as escape characters. This is important for regular expressions where backslashes are common.

\b: In regular expressions, \b represents a word boundary. It matches the position between a word character (like letters and numbers) and a non-word character (like spaces or punctuation).

in: The literal string we're searching for.

\b: Another word boundary to ensure that "in" is matched as a whole word.​
Go MK Websites
+1
Developmentality
+1
Treehouse
+3
regular-expressions.info
+3
Bomberbot
+3

So, \bin\b ensures that we're matching the word "in" only when it appears as a standalone word, not as part of another word like "inside" or "beginning".

case=False: This parameter makes the search case-insensitive, so it will match "in", "In", "IN", etc.​

regex=True: Specifies that the pattern provided is a regular expression. This is necessary because we're using regex-specific syntax like \b.​
Developmentality
+6
Python Tutorial
+6
python.pages.doc.ic.ac.uk
+6

df[ ... ]: Filters the original DataFrame to include only the rows where the condition inside is True.​

shows_with_in =: Assigns the filtered DataFrame to a new variable named shows_with_in.